src/Formatter/ProtobufFormatter.php

```diff
@@ -0,0 +1,352 @@
+<?php
+
+namespace Crell\Serde\Formatter;
+
+use Crell\Serde\Attributes\Binary;
+use Crell\Serde\Attributes\ClassSettings;
+use Crell\Serde\Attributes\Field;
+use Crell\Serde\Attributes\Fixed32;
+use Crell\Serde\Attributes\Fixed64;
+use Crell\Serde\Attributes\Float32;
+use Crell\Serde\Attributes\Float64;
+use Crell\Serde\Attributes\Int32;
+use Crell\Serde\Attributes\Int64;
+use Crell\Serde\Attributes\SFixed32;
+use Crell\Serde\Attributes\SFixed64;
+use Crell\Serde\Attributes\SInt32;
+use Crell\Serde\Attributes\SInt64;
+use Crell\Serde\Attributes\UInt32;
+use Crell\Serde\Attributes\UInt64;
+use Crell\Serde\Dict;
+use Crell\Serde\Formatter\Protobuf\ProtoWireType;
+use Crell\Serde\Formatter\Protobuf\RunningValue;
+use Crell\Serde\Sequence;
+use Crell\Serde\Serializer;
+
+class ProtobufFormatter implements Formatter
+{
+    private const FULL32 = 0xffffffff;
+    private const INT32_MAX = '4294967296';
+    private static Field|null $root = null;
+
+    public function format(): string
+    {
+        return 'protobuf';
+    }
+
+    public function rootField(Serializer $serializer, string $type): Field
+    {
+        return self::$root ??= Field::create('root', $type);
+    }
+
+    public function serializeInitialize(ClassSettings $classDef, Field $rootField): RunningValue
+    {
+        return new RunningValue();
+    }
+
+    public function serializeFinalize(mixed $runningValue, ClassSettings $classDef): string
+    {
+        assert($runningValue instanceof RunningValue);
+        return $runningValue->bytes;
+    }
+
+    public function serializeFloat(mixed $runningValue, Field $field, ?float $next): RunningValue
+    {
+        assert($runningValue instanceof RunningValue);
+        if ($next === null) {
+            $runningValue->skipTag();
+            return $runningValue;
+        }
+
+        $typeField = $field->typeField === null ? null : get_class($field->typeField);
+        switch ($typeField) {
+            case null:
+            case Float64::class:
+                $bytes = pack('e', $next);
+                $runningValue->appendTag(ProtoWireType::Double, $field->fieldNum);
+                $runningValue->appendBytes($bytes);
+                break;
+            case Float32::class:
+                $bytes = pack('g', $next);
+                $runningValue->appendTag(ProtoWireType::Single, $field->fieldNum);
+                $runningValue->appendBytes($bytes);
+                break;
+        }
+
+        return $runningValue;
+    }
+
+    public function serializeString(mixed $runningValue, Field $field, ?string $next): mixed
```

Comment:
There's enough going on here that the switch cases should be split up to their own private methods.  (Which would probably enable using a `match()` instead.)